doctr/models/detection/linknet/pytorch.py

```diff
@@ -174,17 +178,26 @@ def compute_loss(
         seg_target, seg_mask = torch.from_numpy(seg_target).to(dtype=out_map.dtype), torch.from_numpy(seg_mask)
         seg_target, seg_mask = seg_target.to(out_map.device), seg_mask.to(out_map.device)
 
-        # BCE loss
+        # Focal loss
         bce_loss = F.binary_cross_entropy_with_logits(out_map, seg_target, reduction='none')
+        if gamma and gamma < 0:
+            raise ValueError("Value of gamma should be greater than or equal to zero.")
+        # Convert logits to prob, compute gamma factor
+        pred_prob = torch.sigmoid(out_map)[seg_mask]
+        p_t = (seg_target[seg_mask] * pred_prob) + ((1 - seg_target[seg_mask]) * (1 - pred_prob))
+        # Compute alpha factor
+        alpha_factor = seg_target[seg_mask] * alpha + (1 - seg_target[seg_mask]) * (1 - alpha)
+        # compute the final loss
+        focal_loss = (alpha_factor * (1. - p_t) ** gamma * bce_loss[seg_mask]).mean()
```

Comment:
I think we need to address the masking + reduction problem of the dice loss first: once masked, this reduces the tensor to something in 1D. So if one class has 10 times more masked region than another, this will be a problem

I'd suggest doing the following: changing
```python
seg_target[mask].mean()
```
to
```python
mask = mask.to(dtype=torch.foat32)
# Average on N, H, W
class_loss = (seg_target * mask).sum((0, 2, 3)) / mask.sum((0, 2, 3))
loss = class_loss.mean()
```

or average it on H, W only before the final mean

Reply:
I am not sure I understand well here: do you want to remove each ...[mask] occurrence ?

Reply:
So the difference is the following:
- `my_tensor[mask]` is 1D tensor with a number of elements = number of `True` in the mask
- `my_tensor * mask.to(dtype=torch.float32)` has the shape shape as `my_tensor`, it only puts zero on elements that are masked out

Now if you perform a reduction operation like mean:
- in the first case, you divide by the number of elements in `my_tensor[mask]` = `mask.sum()`
- in the second one, you dive by the number of elements in `my_tensor`

And this extends to dimension-specific operations, so if we mask it, we lose the separation of the dimensions to get a contiguous tensor in the end. To properly scale the loss, in the first case, this widely increases the contribution of classes with the highest amount of positive mask (makes no difference if there is only a single class). And since we specifically want to help balance the less-frequent classes here, I suggest leveraging that second option with my suggestion above :+1:

Reply:
(and we need to do this for both the dice loss and the focal loss)

it will only make a difference in cases of multi-class and where the mask isn't only `True` but that might be safer!

Either way, I think we should run a training with the configuration to make sure this yields a positive change :+1: